service/history/api/get_workflow_util.go

```diff
@@ -82,7 +81,18 @@ func GetOrPollMutableState(
 	if request.CurrentBranchToken == nil {
 		request.CurrentBranchToken = response.CurrentBranchToken
 	}
-	if !bytes.Equal(request.CurrentBranchToken, response.CurrentBranchToken) {
+	currentVersionHistory, err := versionhistory.GetCurrentVersionHistory(response.GetVersionHistories())
```

Comment:
i would change the logic from using current branch token + expected event ID to current last version history item
* for getting history asc / desc (not long poll)
  * frontend will only invoke this api once, so even if ms changed, frontend will not notice
  * frontend will return an snapshot of history events
* for long poll waiting new history events
  * this feature only works if caller want to follow the events in asc order
  * branch token (pagination) should contain the current version history last item
    * make sure this last item belongs to current version history
    * update this last item during pagination

Reply:
Let's sync offline about the branch token. Branch token bytes could be changed.